game/view/dialoguebox.lua

```diff
@@ -0,0 +1,388 @@
+        local pause_amount = tonumber(effect_value)
+        if pause_amount then
+          --Increase start-up time in case pause is in the beginning
+          if i == 1 then
+            self.text_start_up_time = self.text_start_up_time + pause_amount
+          else
+            parsed[i-1].time = parsed[i-1].time + pause_amount
+          end
+        else
+          err = true
+        end
+      elseif effect_type == "endl" then
+        local lines_amount = tonumber(effect_value)
+        if lines_amount then
+          y = y + lines_amount * self.text_line_h
+          x = _TEXT_MARGIN
+        else
+          err = true
+        end
+      else
+        err = true
+      end
+
+      --Check for errors
+      if err then
+        error([[Effect invalid!
+             Type = "]]..effect_type..[["
+             Value = "]]..effect_value..[["]])
+      end
+
+    --Common character
+    else
+      local w = font:getWidth(char)
+      local h = font:getHeight(char)
+      local ty = y + self.text_line_h/2 - h/2
+      parsed[i] = {
+        char = char,
+        x = x,
+        y = ty,
+        time = time,
+        color = color,
+        opacity = opacity,
+        style = style,
+        font = font
+      }
+      x = x + w
+
+      --Wrap words
+      if x > _MAX_WIDTH - 2*_TEXT_MARGIN then
+        y = y + self.text_line_h
+        x = _TEXT_MARGIN
+        --Find start of current word
+        local j = i
+        while j >= 1 do
+          if parsed[j].char == " " then break end
+          j = j - 1
+        end
+        if j == 0 then error("Word is too damn big") end
+        --Fix position of every character
+        for k = j+1, i do
+          local w = parsed[k].font:getWidth(parsed[k].char)
+          local h = parsed[k].font:getHeight(parsed[k].char)
+          parsed[k].x = x
+          parsed[k].y = y + self.text_line_h/2 - h/2
+          x = x + w
+        end
+      end
+
+      --Update iterator
+      i = i + 1
+    end
+  end
+
+  return parsed
+end
+
+--Local functions
+
+--Gets a tag effect that starts from given position, and removes that tag from the text
+--A tag must follow the pattern [type_of_effect=value]
+function getTag(text, tag_start_pos)
+  if text:sub(tag_start_pos,tag_start_pos) ~= "[" then
+    error("isn't a valid tag position")
+  end
+
+  --Find tag end position
+  local i = tag_start_pos + 1
+  while i <= text:len() and text:sub(i,i) ~= ']' do
+    i = i + 1
+  end
+  if i > text:len() then
+    error("Reached end of text and tag wasn't closed")
+  end
+
+  --Get complete effect from tag
+  local tag_end_pos = i
+  local effect = text:sub(tag_start_pos + 1, tag_end_pos - 1)
+
+  --Extract type and value from effect
+  local type, value = effect:match("(%w+)=(%w+)")
+  if not type or not value then
+    error("tag didn't match [type=value] pattern")
+  end
+
+  --Remove tag from text
+  text = text:sub(0, tag_start_pos - 1) .. text:sub(tag_end_pos + 1, -1)
+
+
+  return text, type, value
+end
```

Comment:
I recommend separating the parser from the interpreter here. It seems this DialogueBox class does both.

Reply:
Also avoid multiple returns. Prefer to return a table of data instead. It's more innefficient, but at the cost of maintainability

Reply:
Its just a parser, as the name implies. It only checks if the given value is a valid one. The arguments returned are interpreted elsewhere.

Reply:
I'll change multiple returns to be a single table.

Reply:
The logic will be harder to parse and interpretate seperatly, the way the characters in the dialogue box are being stores and used. I'll leave the function as it is for now.